bpo-33012: Fix invalid function cast warnings with gcc 8 for METH_NOARGS

siddhesh opened the pull request:
gcc 8 has added a new warning heuristic to detect invalid function casts and a stock python build seems to hit that warning quite often.  The most common is the cast of a METH_NOARGS function (that uses just one argument) to a PyCFunction.  Fix this by adding a dummy argument to all the affected callback functions.

I have split the changes by directory and also made a separate commit where the change was not just a trivial dummy argument addition, as is the case in a4ee3ad and 58eaf31.

<!-- issue-number: bpo-33012 -->
https://bugs.python.org/issue33012
<!-- /issue-number -->

siddhesh committed:
Objects: Add an unused argument to METH_NOARGS functions

METH_NOARGS functions need only a single argument but they are cast
into a PyCFunction, which takes two arguments.  This triggers an
invalid function cast warning in gcc8 (unreleased as of now) due to
the argument mismatch.  Fix this by adding a dummy unused argument.

siddhesh committed:
Python: Add an unused argument to METH_NOARGS functions

METH_NOARGS functions need only a single argument but they are cast
into a PyCFunction, which takes two arguments.  This triggers an
invalid function cast warning in gcc8 (unreleased as of now) due to
the argument mismatch.  Fix this by adding a dummy unused argument.

siddhesh committed:
Modules: Add an unused argument to METH_NOARGS functions

METH_NOARGS functions need only a single argument but they are cast
into a PyCFunction, which takes two arguments.  This triggers an
invalid function cast warning in gcc8 (unreleased as of now) due to
the argument mismatch.  Fix this by adding a dummy unused argument.

siddhesh committed:
PC: Add or mark an unused argument to METH_NOARGS functions

METH_NOARGS functions need only a single argument but they are cast
into a PyCFunction, which takes two arguments.  This triggers an
invalid function cast warning in gcc8 (unreleased as of now) due to
the argument mismatch.  Fix this by adding a dummy unused argument
wherever it isn't present and mark it as unused to make it consistent
with other similar uses.

siddhesh committed:
Add NEWS item for the internal METH_NOARGS changes

serhiy-storchaka commented on Doc/extending/newtypes.rst:
```diff
@@ -464,8 +464,9 @@ The method is implemented as a C function that takes a :class:`Noddy` (or
-a positional argument tuple or keyword argument dictionary. This method is
-equivalent to the Python method::
+a positional argument tuple or keyword argument dictionary. We add an unused
+argument to allow for a safe cast to a PyCFunction. This method is equivalent to
```
```
:c:type:`PyCFunction`
```

serhiy-storchaka commented on Objects/odictobject.c:
```diff
@@ -1372,8 +1378,8 @@ static PyMethodDef odict_methods[] = {
-    {"__delitem__",     (PyCFunction)odict_mp_ass_sub,  METH_NOARGS,
-     odict_delitem__doc__},
+    {"__delitem__",     (PyCFunction)odict_mp_ass_sub,
+     METH_VARARGS | METH_KEYWORDS, odict_delitem__doc__},
```
This looks incorrect. But an existing code looks incorrect too. I suggest to revert this change (and all other questionable changes) and left them to a separate issue -- [bpo-33031](https://bugs.python.org/issue33031).

siddhesh replied:
OK, dropped from this PR.

serhiy-storchaka commented on Objects/odictobject.c:
```diff
@@ -1466,7 +1472,7 @@ odict_dealloc(PyODictObject *self)
 
 static PyObject *
-odict_repr(PyODictObject *self)
+odict_repr(PyODictObject *self, PyObject *Py_UNUSED(ignored))
```
It should be compatible with `reprfunc`, not `PyCFunction`.

siddhesh replied:
I'll drop this one too since it looks like it will need an additional wrapper like odict_init.

serhiy-storchaka commented on Objects/odictobject.c:
```diff
@@ -1609,15 +1615,15 @@ odict_richcompare(PyObject *v, PyObject *w, int op)
 
 static PyObject *
-odict_iter(PyODictObject *od)
+odict_iter(PyODictObject *od, PyObject *Py_UNUSED(ignored))
```
It should be compatible with `getiterfunc`, not `PyCFunction`.

siddhesh replied:
Likewise, I'll deal with it separately.

serhiy-storchaka commented on Modules/_collectionsmodule.c:
```diff
@@ -1532,7 +1532,7 @@ deque_bool(dequeobject *deque)
 
 static PyObject *
-deque_get_maxlen(dequeobject *deque)
+deque_get_maxlen(dequeobject *deque, PyObject *Py_UNUSED(ignored))
```
It is `getter`.
```
void *Py_UNUSED(ignored)
``

siddhesh replied:
I'm tackling getters in a separate patch, so I'll drop this.

serhiy-storchaka commented on Objects/descrobject.c:
```diff
@@ -439,7 +439,7 @@ calculate_qualname(PyDescrObject *descr)
 
 static PyObject *
-descr_get_qualname(PyDescrObject *descr)
+descr_get_qualname(PyDescrObject *descr, PyObject *Py_UNUSED(ignored))
```
```
void *Py_UNUSED(ignored)
```

siddhesh replied:
Likewise, I've got getter changes in another branch, so I'll fix it there.

serhiy-storchaka commented on Objects/descrobject.c:
```diff
@@ -1139,7 +1139,7 @@ wrapper_text_signature(wrapperobject *wp, void *closure)
 static PyObject *
 wrapper_qualname(wrapperobject *wp)
```
Needs the second argument of type `void *`. As well as other getters.

siddhesh replied:
I've made a separate bug report for the getters: issue33029.

serhiy-storchaka commented on Objects/longobject.c:
```diff
@@ -4554,7 +4554,7 @@ long_or(PyObject *a, PyObject *b)
 
 static PyObject *
-long_long(PyObject *v)
+long_long(PyObject *v, PyObject *Py_UNUSED(ignored))
```
It is used both as `PyCFunction` and `getter`. It would be safer to declare the second parameter as `void *`.

siddhesh replied:
Done.

serhiy-storchaka commented on Objects/odictobject.c:
```diff
@@ -887,14 +887,20 @@ PyDoc_STRVAR(odict_init__doc__,
 /* forward */
-static int odict_init(PyObject *self, PyObject *args, PyObject *kwds);
+static PyObject *
+odict_init(PyObject *self, PyObject *Py_UNUSED(ignored))
```
I suggest to left this for a separate issue.

siddhesh replied:
OK, I'll drop this and the other patch from this PR and make a different one.

serhiy-storchaka commented on Modules/_collectionsmodule.c:
```diff
@@ -512,7 +512,7 @@ deque_inplace_concat(dequeobject *deque, PyObject *other)
 
 static PyObject *
-deque_copy(PyObject *deque)
+deque_copy(PyObject *deque, PyObject *Py_UNUSED(ignored))
```
Explicit cast to `PyCFunction` at line 1590 is not needed.

The same with other functions which exactly match the signature of `PyCFunction`.

siddhesh replied:
I've kept the casts in place because we will eventually need it.  To fully fix the warnings, the `ml_meth` type should be like so:

`PyObject *(*) (PyObject *, PyObject *, ...)`

so that it matches with all of the allowed method types provided that there is an explicit cast in place.  Right now I am not sure if I should change the `PyCFunction` type to that or to make a new `PyCFunctionAny` just for `ml_meth`.  Since we're on this topic, what do you think is the preferred change?  I suppose both change API, but not in a completely incompatible way.

serhiy-storchaka replied:
The casts are not needed if the signature matches `PyCFunction`. It is still needed if the function has type `PyCFunctionWithKeywords`, `PyNoArgsFunction`, etc. Removing unneeded casts will clean up the code and can prevent introducing new bugs when the signature be changed unintentionally.

siddhesh reviewed (commented):
Thanks for the review, I'm pushing a version with most of the fixes you suggested, except for removal of the cast. I've kept the casts in place because we will eventually need it. To fully fix the warnings, the `ml_meth` type should be like so:

`PyObject *(*) (PyObject *, PyObject *, ...)`

so that it matches with all of the allowed method types provided that there is an explicit cast in place. Right now I am not sure if I should change the `PyCFunction` type to that or to make a new `PyCFunctionAny` just for `ml_meth`. Since we're on this topic, what do you think is the preferred change? I suppose both change API, but not in a completely incompatible way.

siddhesh committed:
Doc: Add an unused argument to METH_NOARGS functions

METH_NOARGS functions need only a single argument but they are cast
into a PyCFunction, which takes two arguments.  This triggers an
invalid function cast warning in gcc8 (unreleased as of now) due to
the argument mismatch.  Fix this by adding a dummy unused argument.

serhiy-storchaka commented on PC/bdist_wininst/install.c:
```diff
@@ -631,7 +631,7 @@ static PyObject *PyMessageBox(PyObject *self, PyObject *args)
 }
 
-static PyObject *GetRootHKey(PyObject *self)
+static PyObject *GetRootHKey(PyObject *self, PyObject *Py_UNUSED(ignored))
```
Casting to `PyCFunction` below no longer needed. And this is the only use of `PyCFunction`, thus its declaration can be removed.

siddhesh replied:
Ah I missed this one, I'll fix it.
